### PR TITLE
Add IP address ban link to administration

### PR DESCRIPTION
**Pull request by humble-ahitofel**
Requires the 'ban' module to be enabled.

### CODE REVIEW

File: asklib.install [hunk 1]
```diff
@@ -34,3 +36,13 @@ function asklib_install() {
     ],
   ])->save();
 }
+
+function asklib_update_8001() {
+  
+  $storage_definition = BaseFieldDefinition::create('string')
+    ->setLabel(new TranslatableMarkup('Asker\'s IP address'))
+    ->setDescription(new TranslatableMarkup('IP address where the question was originally made.'));
```
Comment (ajsalminen) [inline asklib.install]:
```suggestion
    ->setDescription(new TranslatableMarkup('IP address from which the question was originally submitted.'));
```

File: src/Entity/Question.php [hunk 2]
```diff
@@ -753,6 +773,10 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
         'weight' => 100,
       ]);
 
+    $fields['ip_address'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Asker\'s IP address'))
+      ->setDescription(t('IP address where the question was originally made.'));
```
Comment (ajsalminen) [inline src/Entity/Question.php]:
```suggestion
      ->setDescription(t('IP address from which the question was originally submitted.'));
```

File: src/Entity/Question.php [hunk 3]
```diff
@@ -49,6 +49,7 @@
  *     "revision" = "vid",
  *     "status" = "published",
  *     "published" = "published",
+ *     "ip_address" = "ip_address"
```
Comment (ajsalminen) [inline src/Entity/Question.php]:
```suggestion
 *     "ip_address" = "ip_address",
```

File: src/Entity/Question.php [hunk 4]
```diff
@@ -381,6 +390,17 @@ public function preSave(EntityStorageInterface $storage) {
     }
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public static function preCreate(EntityStorageInterface $storage, array &$values) {
+    $asker_ip_address = \Drupal::request()->getClientIP();
+    if(!in_array($asker_ip_address, ['::1', '127.0.0.1', '0.0.0.0'])) {
```
Comment (ajsalminen) [inline src/Entity/Question.php]: This seems like a good idea but filter_var() with FILTER_VALIDATE_IP would probably be a better way to do this and it is already possibly done for the value getClientIP() returns. With a quick look I'm not sure what kind of flags are used that check, however.

**Review by ajsalminen — CHANGES_REQUESTED** [review]

File: src/Entity/Question.php (hunk 4 above)
Comment (humble-ahitofel) [reply src/Entity/Question.php]: Added the filter_var(), there was two good rules for filtering out the reserved and private ip addresses. :+1:

**Review by ajsalminen — APPROVED** [review]

**Pull request merged**